File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
from sys import stdin

m, n, k = map(int, stdin.readline().split())
poses = [list(map(int, stdin.readline().split())) for _ in range(k)]
MOVES = [(0, 1), (0, -1), (1, 0), (-1, 0)]

spaces = [[False]*(n+2)] + [[False] + [True]*n + [False] for _ in range(m)] + [[False]*(n+2)]

for lx, ly, rx, ry in poses:
    for row in range(ly + 1, ry + 1):
        for col in range(lx + 1, rx + 1):
            spaces[row][col] = False
            
result = []
for y in range(1, m+1):
    for x in range(1, n+1):
        if not spaces[y][x]:
            continue
        
        area = 0
        stack = [(x, y)]
        spaces[y][x] = False
        while stack:
            px, py = stack.pop()
            area += 1
            for dx, dy in MOVES:
                nx, ny = px + dx, py + dy
                if spaces[ny][nx]:
                    stack.append((nx, ny))
                    spaces[ny][nx] = False
        result.append(area)
        
print(len(result))
print(' '.join(map(str, sorted(result))))

File: 백준/Silver/2583. 영역 구하기/영역 구하기.java
import java.util.*;
import java.io.*;
import java.util.stream.Collectors;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] str = br.readLine().split(" ");
        int m = toInt(str[0]), n = toInt(str[1]), k = toInt(str[2]);
        int[][] poses = new int[k][4];
        for (int i = 0; i < k; i++) {
            String[] pos = br.readLine().split(" ");
            poses[i] = new int[]{toInt(pos[0]), toInt(pos[1]), toInt(pos[2]), toInt(pos[3])};
        }

        final int[][] MOVES = new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        boolean[][] spaces = new boolean[m + 2][n + 2];

        for (int i = 1; i < m + 1; i++) {
            Arrays.fill(spaces[i], 1, n + 1, true);
        }

        for (int[] pos : poses) {
            for (int row = pos[1] + 1; row < pos[3] + 1; row++) {
                for (int col = pos[0] + 1; col < pos[2] + 1; col++) {
                    spaces[row][col] = false;
                }
            }
        }

        List<Integer> result = new ArrayList<>();
        for (int y = 1; y < m + 1; y++) {
            for (int x = 1; x < n + 1; x++) {
                if (!spaces[y][x]) {
                    continue;
                }

                int area = 0;
                Deque<int[]> stack = new ArrayDeque<>();
                stack.offer(new int[]{x, y});
                spaces[y][x] = false;
                while (!stack.isEmpty()) {
                    int[] ppos = stack.poll();
                    area++;
                    for (int[] dpos : MOVES) {
                        int nx = ppos[0] + dpos[0], ny = ppos[1] + dpos[1];
                        if (spaces[ny][nx]) {
                            stack.offer(new int[]{nx, ny});
                            spaces[ny][nx] = false;
                        }
                    }
                }
                result.add(area);
            }
        }

        System.out.println(result.size());
        Collections.sort(result);
        System.out.println(result.stream().map(String::valueOf).collect(Collectors.joining(" ")));
    }

    public static int toInt(String s) {
        return Integer.parseInt(s);
    }
}